File: src/tools/discovery.ts
/**
 * get_join_paths tool handler.
 * Discovers and returns JOIN paths between two tables.
 */

import type { CatalogStore } from "../catalog/store.js";
import type { Embedder } from "../semantic/embedder.js";
import { discoverJoinPaths, computeSchemaHash, discoverCrossSourceJoinPaths, computeCrossSourceSchemaHash, CROSS_SOURCE_ID } from "../discovery/joins.js";
import { discoverBusinessDomains } from "../discovery/domains.js";

/**
 * Handle get_join_paths request.
 * Checks cache, re-discovers if dirty, returns direct + indirect paths.
 */
export async function handleGetJoinPaths(
  params: Record<string, unknown>,
  catalog: CatalogStore,
  embedder?: Embedder | null,
): Promise<string> {
  const tableA = params.table_a as string;
  const tableB = params.table_b as string;

  if (!tableA || !tableB) {
    throw new Error("Both table_a and table_b are required");
  }

  // Find which source(s) contain these tables
  const allTables = catalog.getTables();
  const sourcesForA = new Set(allTables.filter((t) => t.table_name === tableA).map((t) => t.source_id));
  const sourcesForB = new Set(allTables.filter((t) => t.table_name === tableB).map((t) => t.source_id));

  // Validate both tables exist
  if (sourcesForA.size === 0) {
    throw new Error(`Table "${tableA}" not found in any connected source`);
  }
  if (sourcesForB.size === 0) {
    throw new Error(`Table "${tableB}" not found in any connected source`);
  }

  // Find common sources or all relevant sources
  const relevantSources = new Set<string>();
  for (const s of sourcesForA) {
    if (sourcesForB.has(s)) relevantSources.add(s);
  }
  // If no common source, include all sources (cross-source join)
  if (relevantSources.size === 0) {
    for (const s of sourcesForA) relevantSources.add(s);
    for (const s of sourcesForB) relevantSources.add(s);
  }

  // For each relevant source, check cache and discover if needed (within-source paths)
  for (const sourceId of relevantSources) {
    const meta = catalog.getCacheMeta(sourceId);
    const currentHash = computeSchemaHash(catalog, sourceId);

    const needsRefresh = !meta || meta.dirty === 1 || meta.schemaHash !== currentHash;

    if (needsRefresh) {
      catalog.clearJoinPaths(sourceId);
      const paths = await discoverJoinPaths(catalog, sourceId, embedder);
      for (const p of paths) {
        catalog.saveJoinPath(p);
      }
      catalog.setCacheMeta(sourceId, currentHash);
    }
  }

  // Cross-source discovery: check cache and discover if needed
  {
    const crossMeta = catalog.getCacheMeta(CROSS_SOURCE_ID);
    const crossHash = computeCrossSourceSchemaHash(catalog);
    const needsCrossRefresh = !crossMeta || crossMeta.dirty === 1 || crossMeta.schemaHash !== crossHash;

    if (needsCrossRefresh) {
      catalog.clearJoinPaths(CROSS_SOURCE_ID);
      const crossPaths = await discoverCrossSourceJoinPaths(catalog, embedder);
      for (const p of crossPaths) {
        catalog.saveJoinPath(p);
      }
      catalog.setCacheMeta(CROSS_SOURCE_ID, crossHash);
    }
  }

  // Fetch paths for the requested table pair (includes both within-source and cross-source)
  const paths = catalog.getJoinPaths(tableA, tableB);

  const directPaths = paths.filter((p) => !p.signalSource.startsWith("indirect:"));
  const indirectPaths = paths.filter((p) => p.signalSource.startsWith("indirect:"));

  const hasLowConfidence = paths.some((p) => p.confidence < 0.6);
  const warning = hasLowConfidence ? "low confidence, manual verification recommended" : null;

  return JSON.stringify({
    direct_paths: directPaths.map((p) => ({
      join_sql: p.sqlTemplate,
      confidence: Math.round(p.confidence * 100) / 100,
      signal: p.signalSource,
      join_condition: p.joinCondition,
    })),
    indirect_paths: indirectPaths.map((p) => ({
      via: p.signalSource.replace("indirect:via_", ""),
      join_sql: p.sqlTemplate,
      confidence: Math.round(p.confidence * 100) / 100,
      signal: p.signalSource,
      join_condition: p.joinCondition,
    })),
    warning,
  }, null, 2);
}

/**
 * Handle get_business_domains request.
 * Checks cache, re-discovers if dirty, returns clustered domains.
 */
export async function handleGetBusinessDomains(
  params: Record<string, unknown>,
  catalog: CatalogStore,
  embedder?: Embedder | null,
): Promise<string> {
  const datasource = params.datasource as string | undefined;

  // Determine which sources need refresh
  const sources = datasource
    ? catalog.getSources().filter((s) => s.id === datasource)
    : catalog.getSources();

  if (sources.length === 0 && datasource) {
    throw new Error(`Data source "${datasource}" not found`);
  }

  // Check if any source is dirty or has changed schema
  let needsRefresh = false;
  for (const source of sources) {
    const meta = catalog.getCacheMeta(source.id);
    const currentHash = computeSchemaHash(catalog, source.id);
    if (!meta || meta.dirty === 1 || meta.schemaHash !== currentHash) {
      needsRefresh = true;
      break;
    }
  }

  // Also refresh if no domains exist yet
  if (!needsRefresh && catalog.getDomains().length === 0) {
    needsRefresh = true;
  }

  if (needsRefresh) {
    catalog.clearDomains();
    const domains = await discoverBusinessDomains(catalog, embedder);
    for (const d of domains) {
      catalog.saveDomain(d);
    }
    // Update cache for all sources
    for (const source of sources) {
      const hash = computeSchemaHash(catalog, source.id);
      catalog.setCacheMeta(source.id, hash);
    }
  }

  const domains = catalog.getDomains();

  // Filter by datasource if specified
  let filteredDomains = domains;
  if (datasource) {
    const sourceTables = new Set(
      catalog.getTables(datasource).map((t) => t.table_name),
    );
    filteredDomains = domains
      .map((d) => ({
        ...d,
        tableNames: d.tableNames.filter((t) => sourceTables.has(t)),
      }))
      .filter((d) => d.tableNames.length > 0);
  }

  // Compute unclustered tables
  const allTables = datasource
    ? catalog.getTables(datasource)
    : catalog.getTables();
  const clusteredTables = new Set(filteredDomains.flatMap((d) => d.tableNames));
  const unclustered = allTables
    .map((t) => t.table_name)
    .filter((t) => !clusteredTables.has(t));

  return JSON.stringify({
    domains: filteredDomains.map((d) => ({
      name: d.domainName,
      tables: d.tableNames,
      keywords: d.keywords,
      table_count: d.tableNames.length,
    })),
    total_tables: allTables.length,
    unclustered,
  }, null, 2);
}

File: lucid_skill/tools/discovery.py
"""get_join_paths / get_business_domains tool handlers."""

from __future__ import annotations

import json
from typing import Any

from lucid_skill.catalog.store import CatalogStore
from lucid_skill.discovery.domains import discover_business_domains
from lucid_skill.discovery.joins import (
    CROSS_SOURCE_ID,
    compute_cross_source_schema_hash,
    compute_schema_hash,
    discover_cross_source_join_paths,
    discover_join_paths,
)
from lucid_skill.semantic.embedder import Embedder


def handle_get_join_paths(
    params: dict[str, Any],
    catalog: CatalogStore,
    embedder: Embedder | None = None,
) -> str:
    """Discover and return JOIN paths between two tables.

    Checks cache, re-discovers if dirty, returns direct + indirect paths.
    """
    table_a: str | None = params.get("table_a")
    table_b: str | None = params.get("table_b")

    if not table_a or not table_b:
        raise ValueError("Both table_a and table_b are required")

    # Find which source(s) contain these tables
    all_tables = catalog.get_tables()
    sources_for_a = {t["source_id"] for t in all_tables if t["table_name"] == table_a}
    sources_for_b = {t["source_id"] for t in all_tables if t["table_name"] == table_b}

    # Validate both tables exist
    if not sources_for_a:
        raise ValueError(f'Table "{table_a}" not found in any connected source')
    if not sources_for_b:
        raise ValueError(f'Table "{table_b}" not found in any connected source')

    # Find common sources or all relevant sources
    relevant_sources: set[str] = sources_for_a & sources_for_b
    # If no common source, include all sources (cross-source join)
    if not relevant_sources:
        relevant_sources = sources_for_a | sources_for_b

    # For each relevant source, check cache and discover if needed (within-source paths)
    for source_id in relevant_sources:
        meta = catalog.get_cache_meta(source_id)
        current_hash = compute_schema_hash(catalog, source_id)

        needs_refresh = (
            not meta
            or meta.dirty == 1
            or meta.schema_hash != current_hash
        )

        if needs_refresh:
            catalog.clear_join_paths(source_id)
            paths = discover_join_paths(catalog, source_id, embedder)
            for p in paths:
                catalog.save_join_path(p)
            catalog.set_cache_meta(source_id, current_hash)

    # Cross-source discovery: check cache and discover if needed
    cross_meta = catalog.get_cache_meta(CROSS_SOURCE_ID)
    cross_hash = compute_cross_source_schema_hash(catalog)
    needs_cross_refresh = (
        not cross_meta
        or cross_meta.dirty == 1
        or cross_meta.schema_hash != cross_hash
    )

    if needs_cross_refresh:
        catalog.clear_join_paths(CROSS_SOURCE_ID)
        cross_paths = discover_cross_source_join_paths(catalog, embedder)
        for p in cross_paths:
            catalog.save_join_path(p)
        catalog.set_cache_meta(CROSS_SOURCE_ID, cross_hash)

    # Fetch paths for the requested table pair (includes both within-source and cross-source)
    paths = catalog.get_join_paths(table_a, table_b)

    direct_paths = [p for p in paths if not p.signal_source.startswith("indirect:")]
    indirect_paths = [p for p in paths if p.signal_source.startswith("indirect:")]

    has_low_confidence = any(p.confidence < 0.6 for p in paths)
    warning = "low confidence, manual verification recommended" if has_low_confidence else None

    return json.dumps(
        {
            "direct_paths": [
                {
                    "join_sql": p.sql_template,
                    "confidence": round(p.confidence * 100) / 100,
                    "signal": p.signal_source,
                    "join_condition": p.join_condition,
                }
                for p in direct_paths
            ],
            "indirect_paths": [
                {
                    "via": p.signal_source.replace("indirect:via_", ""),
                    "join_sql": p.sql_template,
                    "confidence": round(p.confidence * 100) / 100,
                    "signal": p.signal_source,
                    "join_condition": p.join_condition,
                }
                for p in indirect_paths
            ],
            "warning": warning,
        },
        indent=2,
    )


def handle_get_business_domains(
    params: dict[str, Any],
    catalog: CatalogStore,
    embedder: Embedder | None = None,
) -> str:
    """Discover and return business domain clusters.

    Checks cache, re-clusters if needed, returns domains.
    """
    datasource: str | None = params.get("datasource")

    # Determine which sources need refresh
    all_sources = catalog.get_sources()
    sources = (
        [s for s in all_sources if s["id"] == datasource]
        if datasource
        else all_sources
    )

    if len(sources) == 0 and datasource:
        raise ValueError(f'Data source "{datasource}" not found')

    # Check if any source is dirty or has changed schema
    needs_refresh = False
    for source in sources:
        meta = catalog.get_cache_meta(source["id"])
        current_hash = compute_schema_hash(catalog, source["id"])
        if not meta or meta.dirty == 1 or meta.schema_hash != current_hash:
            needs_refresh = True
            break

    # Also refresh if no domains exist yet
    if not needs_refresh and len(catalog.get_domains()) == 0:
        needs_refresh = True

    if needs_refresh:
        catalog.clear_domains()
        domains = discover_business_domains(catalog, embedder)
        for d in domains:
            catalog.save_domain(d)
        # Update cache for all sources
        for source in sources:
            hash_val = compute_schema_hash(catalog, source["id"])
            catalog.set_cache_meta(source["id"], hash_val)

    domains = catalog.get_domains()

    # Filter by datasource if specified
    filtered_domains = domains
    if datasource:
        source_tables = {
            t["table_name"] for t in catalog.get_tables(datasource)
        }
        filtered_domains = []
        for d in domains:
            filtered_table_names = [t for t in d.table_names if t in source_tables]
            if filtered_table_names:
                from lucid_skill.types import BusinessDomain

                filtered_domains.append(
                    BusinessDomain(
                        domain_id=d.domain_id,
                        domain_name=d.domain_name,
                        table_names=filtered_table_names,
                        keywords=d.keywords,
                        created_at=d.created_at,
                        version=d.version,
                    )
                )

    # Compute unclustered tables
    all_tables = (
        catalog.get_tables(datasource) if datasource else catalog.get_tables()
    )
    clustered_tables = {t for d in filtered_domains for t in d.table_names}
    unclustered = [
        t["table_name"] for t in all_tables if t["table_name"] not in clustered_tables
    ]

    return json.dumps(
        {
            "domains": [
                {
                    "name": d.domain_name,
                    "tables": d.table_names,
                    "keywords": d.keywords,
                    "table_count": len(d.table_names),
                }
                for d in filtered_domains
            ],
            "total_tables": len(all_tables),
            "unclustered": unclustered,
        },
        indent=2,
    )
